Source/Core/DiscIO/NANDImporter.cpp

```diff
@@ -0,0 +1,164 @@
+  }
+  m_nand_keys = std::vector<u8>(NAND_KEYS_SIZE);
+  file.ReadBytes(m_nand_keys.data(), NAND_KEYS_SIZE);
+  return true;
+}
+
+void NANDImporter::FindSuper()
+{
+  u32 super = 0;
+  u32 newest_version = 0;
+  for (int pos = 0x1fc00000; pos < NAND_SIZE; pos += 0x40000)
+  {
+    if (!memcmp(m_nand.data() + pos, "SFFS", 4))
+    {
+      u32 version = Common::swap32(m_nand.data() + pos + 4);
+      if (super == 0 || version > newest_version)
+      {
+        super = pos;
+        newest_version = version;
+      }
+    }
+  }
+  m_nand_fat_offset = super + 0xC;
+  m_nand_fst_offset = m_nand_fat_offset + 0x10000;
+}
+
+std::string NANDImporter::GetPath(NANDFSTEntry* entry, const std::string& parent_path)
+{
+  std::string name(reinterpret_cast<char*>(entry->name), sizeof(NANDFSTEntry::name));
+  // Get rid of any extra null characters
+  while (name.back() == '\0')
+    name.pop_back();
+
+  if (name.front() == '/' || parent_path.back() == '/')
+    return parent_path + name;
+  else
+    return parent_path + "/" + name;
+}
+
+void NANDImporter::ProcessEntry(u16 entry_number, const std::string& parent_path)
+{
+  NANDFSTEntry* entry = reinterpret_cast<NANDFSTEntry*>(
+      m_nand.data() + m_nand_fst_offset + sizeof(NANDFSTEntry) * Common::swap16(entry_number));
+  if (entry->sib != 0xffff)
+    ProcessEntry(entry->sib, parent_path);
+
+  if ((entry->mode & 3) == 1)
+    ProcessFile(entry, parent_path);
+  else if ((entry->mode & 3) == 2)
+    ProcessDirectory(entry, parent_path);
+}
+
+void NANDImporter::ProcessDirectory(NANDFSTEntry* entry, const std::string& parent_path)
+{
+  const std::string path = GetPath(entry, parent_path);
+  File::CreateDir(path);
+
+  if (entry->sub != 0xFFFF)
+    ProcessEntry(entry->sub, path);
+}
+
+void NANDImporter::ProcessFile(NANDFSTEntry* entry, const std::string& parent_path)
+{
+  const std::string path = GetPath(entry, parent_path);
+  File::IOFile file(path, "wb");
+  std::array<u8, 16> key{};
+  memcpy(key.data(), m_nand_keys.data() + NAND_AES_KEY_OFFSET, key.size());
+  u16 sub = Common::swap16(entry->sub);
+
+  for (u32 remaining_bytes = Common::swap32(entry->size); remaining_bytes > 0;)
+  {
+    std::array<u8, 16> iv{};
+    std::vector<u8> block = Common::AES::Decrypt(
+        key.data(), iv.data(), m_nand.data() + NAND_FAT_BLOCK_SIZE * sub, NAND_FAT_BLOCK_SIZE);
+    u32 size = remaining_bytes < NAND_FAT_BLOCK_SIZE ? remaining_bytes : NAND_FAT_BLOCK_SIZE;
+    file.WriteBytes(block.data(), size);
+    remaining_bytes -= size;
+    sub = Common::swap16(m_nand.data() + m_nand_fat_offset + 2 * sub);
+  }
+}
+
+void NANDImporter::ExtractCertificates(const std::string& nand_root)
+{
+  const std::string title_contents_path =
+      nand_root + "/title/00000001/0000000d/content/00000011.app";
+  File::IOFile file(title_contents_path, "rb");
+  if (!file)
+  {
+    PanicAlertT("Unable to open %s! Refer to "
+                "https://wiki.dolphin-emu.org/index.php?title=Wii_Network_Guide to set up "
```

Comment:
Please use the URL https://dolphin-emu.org/docs/guides/wii-network-guide/ instead. It's what we already use: https://github.com/dolphin-emu/dolphin/blob/7932efc78a9470dbeb262a0a8e95035e63e50f63/Source/Core/Core/IOS/Network/SSL.cpp#L112